src/catalog/dependency_manager.cpp

```diff
@@ -27,166 +115,311 @@ void DependencyManager::AddObject(CatalogTransaction transaction, CatalogEntry &
+	if (!old_connections_p) {
+		// Nothing depends on this object and this object doesn't depend on anything either
+		return;
+	}
+	auto &old_connections = *old_connections_p;
+
+	dependency_set_t preserved_dependents;
+	old_connections.ScanDependents(transaction, [&](DependencyCatalogEntry &dep) {
+		// It makes no sense to have a schema depend on anything
+		D_ASSERT(dep.EntryType() != CatalogType::SCHEMA_ENTRY);
+
+		auto entry = LookupEntry(transaction, dep);
+		if (!entry) {
+			return;
 		}
-		if (dep.dependency_type == DependencyType::DEPENDENCY_OWNS) {
-			// the dependent object is owned by the current object
-			owned_objects_to_add.push_back(dep.entry);
-			continue;
+		if (dep.Type() == DependencyType::DEPENDENCY_OWNS) {
+			preserved_dependents.insert(Dependency(*entry, dep.Type()));
+			return;
 		}
 		// conflict: attempting to alter this object but the dependent object still exists
 		// no cascade and there are objects that depend on this object: throw error
 		throw DependencyException("Cannot alter entry \"%s\" because there are entries that "
 		                          "depend on it.",
 		                          old_obj.name);
-	}
-	// add the new object to the dependents_map of each object that it depends on
-	auto &old_dependencies = dependencies_map[old_obj];
-	for (auto &dep : old_dependencies) {
-		auto &dependency = dep.get();
-		dependents_map[dependency].insert(new_obj);
-	}
+	});
 
-	// We might have to add a type dependency
-	// add the new object to the dependency manager
-	dependents_map[new_obj] = dependency_set_t();
-	dependencies_map[new_obj] = old_dependencies;
+	// Keep old dependencies
+	dependency_set_t dependency_list;
+	old_connections.ScanDependencies(transaction, [&](DependencyCatalogEntry &dep) {
+		auto entry = LookupEntry(transaction, dep);
+		if (!entry) {
+			return;
+		}
+		dependency_list.insert(Dependency(*entry, dep.Type()));
+	});
 
-	for (auto &dependency : owned_objects_to_add) {
-		dependents_map[new_obj].insert(Dependency(dependency, DependencyType::DEPENDENCY_OWNS));
-		dependents_map[dependency].insert(Dependency(new_obj, DependencyType::DEPENDENCY_OWNED_BY));
-		dependencies_map[new_obj].insert(dependency);
+	// FIXME: we should update dependencies in the future
+	// some alters could cause dependencies to change (imagine types of table columns)
+	// or DEFAULT depending on a sequence
+	if (old_obj.name != new_obj.name) {
+		CleanupDependencies(transaction, old_obj);
 	}
-}
 
-void DependencyManager::EraseObject(CatalogEntry &object) {
-	// obtain the writing lock
-	EraseObjectInternal(object);
-}
+	for (auto &dep : dependency_list) {
+		auto &other = dep.entry.get();
+		auto other_connections = GetDependencySet(transaction, other);
+		// Register that the new version of this object still has this dependency.
+		// FIXME: what should the dependency type be???
+		other_connections->AddDependent(transaction, new_obj, DependencyType::DEPENDENCY_REGULAR);
```

Comment:
We should look up the corresponding dependent-entry that `other_connections` holds on `object` before CleanupDependencies and re-add that connection here.